File: lib/std/unicode-inline.js
/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

// combining marks
var _rx_combining = /([\0-\u02FF\u0370-\u1DBF\u1E00-\u20CF\u2100-\uD7FF\uDC00-\uFE1F\uFE30-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF])([\u0300-\u036F\u1DC0-\u1DFF\u20D0-\u20FF\uFE20-\uFE2F]*)|([\s\S])/g;

// normalize a string
function _normalize(s,nf) {
  if (typeof s.normalize === "function") {
    return s.normalize(nf || "NFC");
  }
  else {
    // .. no unicode support on this platform; let's hope for the best
    return s; 
  }
}

// convert to a vector of grapheme strings
function _graphemes(s) {
  var gs = [];
  s.replace(_rx_combining, function(m) {
    gs.push(m);
  });
  return gs;
}

File: lib/std/unicode-inline.cs
/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Text;
using System.Globalization;

static class Unicode {
  public static string Normalize( string s, string norm ) {
    NormalizationForm nf;
    if (norm=="NFD") nf = NormalizationForm.FormD;
    else if (norm=="NFKC") nf = NormalizationForm.FormKC;
    else if (norm=="NFKD") nf = NormalizationForm.FormKD;
    else nf = NormalizationForm.FormC;
    return s.Normalize(nf);
  }
  public static string[] Graphemes( string s ) {
    if (String.IsNullOrEmpty(s)) return new string[0];
    string[] v = new string[s.Length];
    TextElementEnumerator e = StringInfo.GetTextElementEnumerator(s);
    int i = 0;
    while( e.MoveNext() ) {
      v[i] = (string)e.Current;
      i++;
    }
    Array.Resize(ref v,i);
    return v;
  }
}
